src/main/scala/com/advancedtelematic/director/db/DeviceUpdateAssignmentRepository.scala

```diff
@@ -0,0 +1,113 @@
+  import Schema.deviceUpdateAssignments
+
+  def exists(namespace: Namespace, deviceId: DeviceId, version: Int): Future[Boolean] =
+    db.run(existsAction(namespace, deviceId, version))
+
+  protected [db] def existsAction(namespace: Namespace, deviceId: DeviceId, version: Int)
+  : DBIO[Boolean] =
+    deviceUpdateAssignments
+      .filter(_.deviceId === deviceId)
+      .filter(_.version === version)
+      .exists
+      .result
+
+  protected [db] def persistAction(
+    namespace: Namespace,
+    deviceId: DeviceId,
+    correlationId: Option[CorrelationId],
+    updateId: Option[UpdateId],
+    version: Int)
+  : DBIO[DeviceUpdateAssignment] = {
+    val assignment = DeviceUpdateAssignment(namespace, deviceId, correlationId, updateId, version, inFlight = false)
+
+    (deviceUpdateAssignments += assignment)
+      .map(_ => assignment)
+      .handleIntegrityErrors(ConflictingTarget)
+  }
+
+  protected [db] def persistAction(
+    namespace: Namespace,
+    deviceId: DeviceId,
+    correlationId: Option[CorrelationId],
+    updateId: Option[UpdateId])
+  : DBIO[Int] = for {
+    version <- fetchLatestAction(namespace, deviceId).asTry.flatMap {
+      case Success(x) => DBIO.successful(x)
+      case Failure(NoTargetsScheduled) => DBIO.successful(0)
+      case Failure(ex) => DBIO.failed(ex)
+    }
+    new_version = version + 1
+    _ <- persistAction(namespace, deviceId, correlationId, updateId, new_version)
+  } yield new_version
+
+
+  protected [db] def fetchLatestAction(namespace: Namespace, deviceId: DeviceId): DBIO[Int] =
+    deviceUpdateAssignments
+    /**
+      // TODO: Filter by namespace causes deadlock exception for test
+      // "Can schedule several updates for the same device at the same time" in  FileCacheSpec
+      .filter(_.namespace === namespace)
+     */
+      .filter(_.deviceId === deviceId)
+      .map(_.version)
+      .forUpdate
+      .max
+      .result
+      .failIfNone(NoTargetsScheduled)
+
+  protected [db] def fetchAction(namespace: Namespace, deviceId: DeviceId, version: Int)
+  : DBIO[DeviceUpdateAssignment] =
+    deviceUpdateAssignments
+      .filter(_.namespace === namespace)
+      .filter(_.deviceId === deviceId)
+      .filter(_.version === version)
+      .result
+      .failIfMany
+      .failIfNone(NoTargetsScheduled)
+
+
+  protected [db] def fetchAllAfterAction(namespace: Namespace, deviceId: DeviceId, fromVersion: Int)
+  : DBIO[Seq[DeviceUpdateAssignment]] =
+    deviceUpdateAssignments
+      .filter(_.namespace === namespace)
+      .filter(_.deviceId === deviceId)
+      .filter(_.version > fromVersion)
+      .sortBy(_.version)
+      .result
+
+  def getUpdatesFromTo(namespace: Namespace, deviceId: DeviceId, fromVersion: Int, toVersion: Int)
+  : Future[Seq[DeviceUpdateAssignment]] = db.run {
+    Schema.deviceUpdateAssignments
+      .filter(_.namespace === namespace)
+      .filter(_.deviceId === deviceId)
+      .filter(_.version > fromVersion)
```

Comment:
Would it make sense to have the inclusion in the lower bound instead? Or is this actually the use we need for this method? Slicing like this is usually the interval `[from, to)`, e. g. Java `String#substring`, so doing it the other way around might be unexpected to the user of this method.